### PR TITLE
executor: fix IndexLookUpExecutor data race when open/close multiple times

### PR DESCRIPTION
**Pull request by tiancaiamao**
IndexLookUpExecutor may be `Open` and `Close` several times when its parent is an Apply executor.
After `Close`, some background goroutine is still working, so there is data race between the next `Open` call and background goroutine:

```
func (e *IndexLookUpExecutor) Open() error {
	e.finished = make(chan struct{})   // here

func (e *IndexLookUpExecutor) pickAndExecTask() {
        case <-e.finished:                // and here
```

A better way to fix is that `Close` should wait until all background to exit, but so difficult for the messy code.
Here is just a simple compromise: pass `e.finished` as parameter to avoid race.

@hanfei1991 @zimulala @shenli

### CODE REVIEW

File: executor/new_distsql.go [hunk 1]
```diff
@@ -323,7 +323,7 @@ func (e *IndexLookUpExecutor) Open() error {
 	// e.taskChan serves as a pipeline, so fetching index and getting table data can
 	// run concurrently.
 	e.taskChan = make(chan *lookupTableTask, atomic.LoadInt32(&LookupTableTaskChannelSize))
-	go e.fetchHandlesAndStartWorkers()
+	go e.fetchHandlesAndStartWorkers(e.finished)
```
Comment (ngaut) [inline executor/new_distsql.go]: So there is no need to define finished as member of e.

Comment (tiancaiamao) [inline executor/new_distsql.go]: If `finished` is not a member of e, how can we close it?

**Comment by shenli** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM

**Comment by tiancaiamao** [comment]
I'll do a refactor to fix this.

**Pull request closed without merging**